substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_jdk_nashorn_api_scripting_NashornScriptEngineFactory.java

```diff
@@ -28,9 +28,10 @@
 
 import com.oracle.svm.core.annotate.Substitute;
 import com.oracle.svm.core.annotate.TargetClass;
+import com.oracle.svm.core.jdk.NashornSupport.NashornAvailable;
 import com.oracle.svm.core.util.VMError;
 
-@TargetClass(className = "jdk.nashorn.api.scripting.NashornScriptEngineFactory", onlyWith = JDK14OrEarlier.class)
+@TargetClass(className = "jdk.nashorn.api.scripting.NashornScriptEngineFactory", onlyWith = {JDK14OrEarlier.class, NashornAvailable.class})
```

Comment:
I fail to see why we need this check since we know that nashorn is available in JDK <= 14 and we already check for that.
What am I missing?

Reply:
As @jerboaa observed, there are JDK <= 14 where module `jdk.scripting.nashorn` is not per-default available as a module (running `java --list-modules` would not show it).

On our side it's not a problem. Our jlinked Java 11 has module `jdk.scripting.nashorn`.

Reply:
> As @jerboaa observed, there are JDK <= 14 where module `jdk.scripting.nashorn` is not per-default available as a module (i.e. running `java --list-modules` would not show it).

Judging by [@jerboaa's initial patch](https://github.com/oracle/graal/commit/5e4b8fa0c90444115930b09c39b1b999cdd5843f#diff-42623de4b000eeb25f4b626861ebe69fd6fbcb7e9480897d1d5c3ff79cb7eedd) it looks like the assumption was that module `jdk.scripting.nashorn` is indeed available in JDK <= 14

The [`--list-modules` part](https://github.com/oracle/graal/commit/5e4b8fa0c90444115930b09c39b1b999cdd5843f#diff-9e053bec9eca0799896bd471363fa6aa1df70269e454b89dff1ea5fb844a480cR540) was only meant to check for `org.graalvm` modules and not for `jdk.scripting.nashorn`.

I suggest we wait for @jerboaa's input on this before pushing it to the merge queue.

Reply:
> assumption was that module jdk.scripting.nashorn is indeed available in JDK <= 14

There is a difference between available and enabled as I tried to explain above.

Reply:
> I suggest we wait for @jerboaa's input on this before pushing it to the merge queue.

No problem.

Reply:
For that reason it's not sufficient to only go by the JDK versions on the substitutions.

Reply:
OK, I see. I was also able to reproduce the issue with OpenJDK 11.0.12 and 11.0.13. Note, however, that the module appears in `java --list-modules`.

Reply:
@zakkak pure modular application (like e.g. `native-image-agent-library`) would need to specify `requires jdk.scripting.nashorn` in their `module-info.java` to have access to it. Otherwise `jdk.scripting.nashorn` is just not available for them.

Thus if you build an image on the module path (i.e. use `-p` `-m` options instead of `-cp` or `-jar` with native-image) it is not a given that `jdk.scripting.nashorn` will be available.